File: Data Structures & Algorithms/top-k-elements-in-list/submission-5.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> frequencies = new HashMap();
        for(int n : nums){
            frequencies.put(n, frequencies.getOrDefault(n, 0) + 1);
        }
        //1 - 1
        //2 - 2
        //3 - 3
        //0 1 2 3 4 5 6 
        //  1 2 3
        List<Integer>[] buckets = new List[nums.length + 1];
        for(int i = 0; i < buckets.length; i++){
            buckets[i] = new ArrayList();
        }
        for(Map.Entry<Integer, Integer> entry : frequencies.entrySet()){
            buckets[entry.getValue()].add(entry.getKey());
        }
        int[] res = new int[k];
        int idx = 0;
        for(int i = buckets.length-1; i > 0; i--){
            for(int n : buckets[i]){
                res[idx] = n;
                idx++;
                if(idx == k){
                    return res;
                }
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-5.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        i = 0
        print(nums)
        while i < len(nums)-2:
            l = i + 1
            r = len(nums)-1
            while l < r:
                val = nums[i] + nums[l] + nums[r]
                if val < 0:
                    l += 1
                elif val > 0:
                    r -= 1
                else:
                    print("i l r:",i,l,r)
                    res.append([nums[i], nums[l], nums[r]])
                    l += 1
                    while l < r and nums[l] == nums[l-1]:
                        l += 1
            i += 1
            while i < len(nums)-2 and nums[i] == nums[i-1]:
                i += 1
        return res
                    

